internal/auth0/organization/data_source.go

```diff
@@ -0,0 +1,166 @@
+package organization
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/auth0/go-auth0/management"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	internalSchema "github.com/auth0/terraform-provider-auth0/internal/schema"
+)
+
+// NewDataSource will return a new auth0_organization data source.
+func NewDataSource() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: readOrganizationForDataSource,
+		Description: "Data source to retrieve a specific Auth0 organization by `organization_id` or `name`.",
+		Schema:      dataSourceSchema(),
+	}
+}
+
+func dataSourceSchema() map[string]*schema.Schema {
+	dataSourceSchema := internalSchema.TransformResourceToDataSource(NewResource().Schema)
+	dataSourceSchema["organization_id"] = &schema.Schema{
+		Type:         schema.TypeString,
+		Optional:     true,
+		Description:  "The ID of the organization. If not provided, `name` must be set.",
+		AtLeastOneOf: []string{"organization_id", "name"},
+	}
+
+	internalSchema.SetExistingAttributesAsOptional(dataSourceSchema, "name")
+	dataSourceSchema["name"].Description = "The name of the organization. " +
+		"If not provided, `organization_id` must be set. " +
+		"For performance, it is advised to use the `organization_id` as a lookup if possible."
+	dataSourceSchema["name"].AtLeastOneOf = []string{"organization_id", "name"}
+
+	dataSourceSchema["connections"] = &schema.Schema{
+		Type:     schema.TypeSet,
+		Computed: true,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"connection_id": {
+					Type:        schema.TypeString,
+					Computed:    true,
+					Description: "The ID of the enabled connection on the organization.",
+				},
+				"assign_membership_on_login": {
+					Type:     schema.TypeBool,
+					Computed: true,
+					Description: "When `true`, all users that log in with this connection will be " +
+						"automatically granted membership in the organization. When `false`, users must be " +
+						"granted membership in the organization before logging in with this connection.",
+				},
+			},
+		},
+	}
+
+	return dataSourceSchema
+}
+
+func readOrganizationForDataSource(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	api := meta.(*management.Management)
+	var foundOrganization *management.Organization
+	var err error
+
+	organizationID := data.Get("organization_id").(string)
+	if organizationID != "" {
+		foundOrganization, err = api.Organization.Read(organizationID)
+		if err != nil {
+			if mErr, ok := err.(management.Error); ok && mErr.Status() == http.StatusNotFound {
+				data.SetId("")
+				return nil
+			}
+			return diag.FromErr(err)
+		}
+	} else {
+		name := data.Get("name").(string)
```

Comment:
Is it necessary to offer looking up the org by name, besides just by ID? Just curious.

Reply:
Not necessarily, we could consider removing it and allowing only lookup based on ID. But then it would be best to remove the functionality from all other data sources as well.

Reply:
Note that this won't apply to all resources (only those that have unique names). Whereas in others it makes sense. If possible, consider having this for some resources only.

Reply:
Note that if customers actually need this functionality in other resources, it can always be added later.

Reply:
You raise a good point, I'll have to investigate the topic a bit deeper. For now I'll keep this as is in the PR however before we cut a release with all the data sources added we'll take a decision whether to only allow id's for the lookup or id's and names.